src/sdf/include/sdf.h

```diff
@@ -0,0 +1,474 @@
+                                      {-1, -1, -1}};
+  return tetTri1[i];
+}
+
+__host__ __device__ glm::ivec4 Neighbors(int i) {
+  constexpr glm::ivec4 neighbors[14] = {{0, 0, 0, 1},     //
+                                        {1, 0, 0, 0},     //
+                                        {0, 1, 0, 0},     //
+                                        {0, 0, 1, 0},     //
+                                        {-1, 0, 0, 1},    //
+                                        {0, -1, 0, 1},    //
+                                        {0, 0, -1, 1},    //
+                                        {-1, -1, -1, 1},  //
+                                        {-1, 0, 0, 0},    //
+                                        {0, -1, 0, 0},    //
+                                        {0, 0, -1, 0},    //
+                                        {0, -1, -1, 1},   //
+                                        {-1, 0, -1, 1},   //
+                                        {-1, -1, 0, 1}};
+  return neighbors[i];
+}
+
+__host__ __device__ Uint64 SpreadBits3(Uint64 v) {
+  v = v & 0x1fffff;
+  v = (v | v << 32) & 0x1f00000000ffff;
+  v = (v | v << 16) & 0x1f0000ff0000ff;
+  v = (v | v << 8) & 0x100f00f00f00f00f;
+  v = (v | v << 4) & 0x10c30c30c30c30c3;
+  v = (v | v << 2) & 0x1249249249249249;
+  return v;
+}
+
+__host__ __device__ Uint64 SqeezeBits3(Uint64 v) {
+  v = v & 0x1249249249249249;
+  v = (v ^ v >> 2) & 0x10c30c30c30c30c3;
+  v = (v ^ v >> 4) & 0x100f00f00f00f00f;
+  v = (v ^ v >> 8) & 0x1f0000ff0000ff;
+  v = (v ^ v >> 16) & 0x1f00000000ffff;
+  v = (v ^ v >> 32) & 0x1fffff;
+  return v;
+}
+
+// This is a modified 3D MortonCode, where the xyz code is shifted by one bit
+// and the w bit is added as the least significant. This allows 21 bits per x,
+// y, and z channel and 1 for w, filling the 64 bit total.
+__host__ __device__ Uint64 MortonCode(const glm::ivec4& index) {
+  return static_cast<Uint64>(index.w) | (SpreadBits3(index.x) << 1) |
+         (SpreadBits3(index.y) << 2) | (SpreadBits3(index.z) << 3);
+}
+
+__host__ __device__ glm::ivec4 DecodeMorton(Uint64 code) {
+  glm::ivec4 index;
+  index.x = SqeezeBits3(code >> 1);
+  index.y = SqeezeBits3(code >> 2);
+  index.z = SqeezeBits3(code >> 3);
+  index.w = code & 0x1u;
+  return index;
+}
+
+struct GridVert {
+  Uint64 key = kOpen;
+  float distance = NAN;
+  int edgeIndex = -1;
+  int edgeVerts[7] = {-1, -1, -1, -1, -1, -1, -1};
+
+  __host__ __device__ int Inside() const { return distance > 0 ? 1 : -1; }
+};
+
+class HashTableD {
+ public:
+  HashTableD(VecDH<GridVert>& alloc, VecDH<uint32_t>& used, uint32_t step = 127)
+      : step_{step}, table_{alloc}, used_{used} {}
+
+  __host__ __device__ int Size() const { return table_.size(); }
+
+  __host__ __device__ bool Insert(const GridVert& vert) {
+    uint32_t idx = vert.key & (Size() - 1);
+    while (1) {
+      Uint64& key = table_[idx].key;
+      const Uint64 found = AtomicCAS(key, kOpen, vert.key);
```

Comment:
As usual I was focused more on the GPU, where locks aren't much of an option. I'm totally open to recommendations though.

Reply:
You can implement a spin lock with atomic.

Reply:
My impression was that atomics are mostly slow when concurrent access is actually happening; do they have much performance impact otherwise? I would expect very little concurrency on these - they will only happen during a hash collision (which is already rare) that also happens to be concurrent. I made the step large so the nearby threads don't tend to step on each other.

Reply:
Yes, they have performance impact even when there is only 1 thread running these atomic instructions.

Reply:
That's good to know, thanks. I'll leave it this way for now, but you're welcome to optimize it - that will be a good learning experience for me.